esi_leap/objects/lease.py

```diff
@@ -160,47 +160,48 @@ def get_all(cls, filters, context=None):
+                                                     resource_uuid),
+                        external=True):
+            self.verify_time_range(
+                start_time, end_time,
+                updates.get('offer_uuid', None),
+                updates.get('parent_lease_uuid', None),
+                resource_type, resource_uuid)
 
-        with utils.lock(utils.get_resource_lock_name(updates['resource_type'],
-                                                     updates['resource_uuid']),
+            db_lease = self.dbapi.lease_create(updates)
+            self._from_db_object(context, self, db_lease)
+
+    def update(self, updates, context=None):
+        # only allow updates to end_time right now
+        if 'end_time' not in updates:
+            return
+        new_end_time = updates['end_time']
+        with utils.lock(utils.get_resource_lock_name(self.resource_type,
+                                                     self.resource_uuid),
                         external=True):
-            if updates['start_time'] >= updates['end_time']:
+            if self.start_time >= new_end_time:
                 raise exception.InvalidTimeRange(
                     resource='lease',
-                    start_time=str(updates['start_time']),
-                    end_time=str(updates['end_time'])
+                    start_time=str(self.start_time),
+                    end_time=str(new_end_time)
                     )
 
-            # check availability
-            if 'offer_uuid' in updates:
-                # lease is being created from an offer
-                related_offer = offer_obj.Offer.get(updates['offer_uuid'])
-
-                if related_offer.status != statuses.AVAILABLE:
-                    raise exception.OfferNotAvailable(
-                        offer_uuid=related_offer.uuid,
-                        status=related_offer.status)
-
-                related_offer.verify_availability(updates['start_time'],
-                                                  updates['end_time'])
-            elif 'parent_lease_uuid' in updates:
-                # lease is a child of an existing lease
-                parent_lease = Lease.get(updates['parent_lease_uuid'])
-
-                if parent_lease.status != statuses.ACTIVE:
-                    raise exception.LeaseNotActive(
-                        updates['parent_lease_uuid'])
-
-                parent_lease.verify_child_availability(updates['start_time'],
-                                                       updates['end_time'])
-            else:
-                ro = get_resource_object(updates['resource_type'],
-                                         updates['resource_uuid'])
-                ro.verify_availability(updates['start_time'],
-                                       updates['end_time'])
+            # only need to check availabilities if new end time is greater
+            # than previous end time
+            if new_end_time > self.end_time:
```

Comment:
In verify_time_range(), it also checks if new_end_time > self.end_time. Is it a duplicate check here?

Reply:
Ah, the difference is that this check is to see if we even need to call `verify_time_range`.  Let's say lease initially runs from time `A----->B`. If the updated end time is before `B`, there's no need to `verify_time_range` at all, since the new lease time range is within the old one. If it's after time `B`, then we verify from `B---->C`.

Reply:
Yes, but if the updated end time is before `B`, do you need to raise an exception to indicate that the new_end_time is invalid?

Reply:
Ah, nope - you're right, in the update case I technically don't need that check, since it'll never happen. However when I consolidated the time checking code into `verify_time_range`, it seemed strange to leave that out; I figured it was better to have a duplicated check than to leave out a condition if/when some other code calls `verify_time_range` in the future.